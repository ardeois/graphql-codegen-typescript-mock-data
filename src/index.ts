import { printSchema, parse, visit, ASTKindToNode, NamedTypeNode, TypeNode, VisitFn } from 'graphql';
import casual from 'casual';
import { PluginFunction } from '@graphql-codegen/plugin-helpers';
import { pascalCase } from 'pascal-case';

export function toPascalCase(str: string) {
    if (str.charAt(0) === '_') {
        return str.replace(
            /^(_*)(.*)/,
            (_match, underscorePrefix, typeName) => `${underscorePrefix}${pascalCase(typeName || '')}`,
        );
    }

    return pascalCase(str || '');
}

const toMockName = (name: string) => {
    const isVowel = name.match(/^[AEIO]/);
    return isVowel ? `an${name}` : `a${name}`;
};

const hashedString = (value: string) => {
    let hash = 0;
    if (value.length === 0) {
        return hash;
    }
    for (let i = 0; i < value.length; i++) {
        const char = value.charCodeAt(i);
        // eslint-disable-next-line no-bitwise
        hash = (hash << 5) - hash + char;
        // eslint-disable-next-line no-bitwise
        hash = hash & hash; // Convert to 32bit integer
    }
    return hash;
};

const getNamedType = (
    typeName: string,
    fieldName: string,
    types: TypeItem[],
    namedType?: NamedTypeNode,
): string | number | boolean => {
    if (!namedType) {
        return '';
    }

    casual.seed(hashedString(typeName + fieldName));
    const name = namedType.name.value;
    switch (name) {
        case 'String':
            return `'${casual.word}'`;
        case 'Float':
            return Math.round(casual.double(0, 10) * 100) / 100;
        case 'ID':
            return `'${casual.uuid}'`;
        case 'Boolean':
            return casual.boolean;
        case 'Int':
            return casual.integer(0, 9999);
        case 'Date':
            return `'${new Date(casual.unix_time).toISOString()}'`;
        default: {
            const foundType = types.find((enumType: TypeItem) => enumType.name === name);
            if (foundType) {
                switch (foundType.type) {
                    case 'enum': {
                        // It's an enum
                        const value = foundType.values ? foundType.values[0] : '';
                        return `${foundType.name}.${toPascalCase(value)}`;
                    }
                    case 'union':
                        // Return the first union type node.
                        return getNamedType(typeName, fieldName, types, foundType.types && foundType.types[0]);
                    default:
                        throw `foundType is unknown: ${foundType.name}: ${foundType.type}`;
                }
            }
            return `${toMockName(name)}()`;
        }
    }
};

const generateMockValue = (
    typeName: string,
    fieldName: string,
    types: TypeItem[],
    currentType: TypeNode,
): string | number | boolean => {
    switch (currentType.kind) {
        case 'NamedType':
            return getNamedType(typeName, fieldName, types, currentType as NamedTypeNode);
        case 'NonNullType':
            return generateMockValue(typeName, fieldName, types, currentType.type);
        case 'ListType': {
            const value = generateMockValue(typeName, fieldName, types, currentType.type);
            return `[${value}]`;
        }
    }
};

const getMockString = (typeName: string, fields: string, addTypename = false) => {
    const typename = addTypename ? `\n        __typename: '${typeName}',` : '';
    return `
export const ${toMockName(typeName)} = (overrides?: Partial<${typeName}>): ${typeName} => {
    return {${typename}
${fields}
        ...overrides
    };
};`;
};

export interface TypescriptMocksPluginConfig {
    typesFile?: string;
    addTypename?: boolean;
}

interface TypeItem {
    name: string;
    type: string;
    values?: string[];
    types?: ReadonlyArray<NamedTypeNode>;
}

type VisitorType = { [K in keyof ASTKindToNode]?: VisitFn<ASTKindToNode[keyof ASTKindToNode], ASTKindToNode[K]> };

// This plugin was generated with the help of ast explorer.
// https://astexplorer.net
// Paste your graphql schema in it, and you'll be able to see what the `astNode` will look like
export const plugin: PluginFunction<TypescriptMocksPluginConfig> = (schema, documents, config) => {
    const printedSchema = printSchema(schema); // Returns a string representation of the schema
    const astNode = parse(printedSchema); // Transforms the string into ASTNode
    // List of types that are enums
    const types: TypeItem[] = [];
    const visitor: VisitorType = {
        EnumTypeDefinition: (node) => {
            const name = node.name.value;
            if (!types.find((enumType: TypeItem) => enumType.name === name)) {
                types.push({
                    name,
                    type: 'enum',
                    values: node.values ? node.values.map((node) => node.name.value) : [],
                });
            }
        },
        UnionTypeDefinition: (node) => {
            const name = node.name.value;
            if (!types.find((enumType) => enumType.name === name)) {
                types.push({
                    name,
                    type: 'union',
                    types: node.types,
                });
            }
        },
        FieldDefinition: (node) => {
            const fieldName = node.name.value;

            return {
                name: fieldName,
                mockFn: (typeName: string) => {
                    const value = generateMockValue(typeName, fieldName, types, node.type);

                    return `        ${fieldName}: ${value},`;
                },
            };
        },
        InputObjectTypeDefinition: (node) => {
            const fieldName = node.name.value;

            return {
                typeName: fieldName,
                mockFn: () => {
                    const mockFields = node.fields
                        ? node.fields
                              .map((field) => {
                                  const value = generateMockValue(fieldName, field.name.value, types, field.type);

                                  return `        ${field.name.value}: ${value},`;
                              })
                              .join('\n')
                        : '';

                    return getMockString(fieldName, mockFields, false);
                },
            };
        },
        ObjectTypeDefinition: (node) => {
            // This function triggered per each type
            const typeName = node.name.value;

            if (typeName === 'Query' || typeName === 'Mutation') {
                return null;
            }

            const { fields } = node;
            return {
                typeName,
                mockFn: () => {
                    const mockFields = fields ? fields.map(({ mockFn }: any) => mockFn(typeName)).join('\n') : '';

                    return getMockString(typeName, mockFields, !!config.addTypename);
                },
            };
        },
    };

    const result: any = visit(astNode, { leave: visitor });
    const definitions = result.definitions.filter((definition: any) => !!definition);
    const typesFile = config.typesFile ? config.typesFile.replace(/\.[\w]+$/, '') : null;
    const typeImports = definitions
        .map(({ typeName }: { typeName: string }) => typeName)
        .filter((typeName: string) => !!typeName);
    typeImports.push(...types.map(({ name }) => name));
    // List of function that will generate the mock.
    // We generate it after having visited because we need to distinct types from enums
    const mockFns = definitions.map(({ mockFn }: any) => mockFn).filter((mockFn: Function) => !!mockFn);
    const typesFileImport = typesFile
        ? `/* eslint-disable @typescript-eslint/no-use-before-define,@typescript-eslint/no-unused-vars */
import { ${typeImports.join(', ')} } from '${typesFile}';\n`
        : '';

    return `${typesFileImport}${mockFns.map((mockFn: Function) => mockFn()).join('\n')}
`;
};
