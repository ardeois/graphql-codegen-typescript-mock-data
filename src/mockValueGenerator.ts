import { faker } from '@faker-js/faker';
import casual from 'casual';

interface MockValueGenerator {
    dynamicValues: boolean;
    word: () => string;
    uuid: () => string;
    boolean: () => boolean | string;
    integer: () => number | string;
    float: () => number | string;
    date: () => string;
    seed: (seed: number) => void;
}

type MockValueGeneratorOptions = {
    dynamicValues: boolean;
};

type FunctionTokens = Record<'import' | 'seed' | 'seedFunction', string>;

type SetupMockValueGeneratorOptions = {
    generateLibrary: 'casual' | 'faker';
    dynamicValues: boolean;
};

class CasualMockValueGenerator implements MockValueGenerator {
    dynamicValues: boolean;

    constructor(opts: MockValueGeneratorOptions) {
        this.dynamicValues = opts.dynamicValues;
    }

    word = () => (this.dynamicValues ? `casual.word` : `'${casual.word}'`);
    uuid = () => (this.dynamicValues ? `casual.uuid` : `'${casual.uuid}'`);
    boolean = () => (this.dynamicValues ? `casual.boolean` : casual.boolean);
    integer = () => (this.dynamicValues ? `casual.integer(0, 9999)` : `${casual.integer(0, 9999)}`);
    float = () =>
        this.dynamicValues
            ? `Math.round(casual.double(0, 10) * 100) / 100`
            : `${Math.round(casual.double(0, 10) * 100) / 100}`;
    date = () =>
        this.dynamicValues
            ? `new Date(casual.unix_time).toISOString()`
            : `'${new Date(casual.unix_time).toISOString()}'`;
    seed = (seed: number) => casual.seed(seed);
}

const casualFunctionTokens: FunctionTokens = {
    import: `import casual from 'casual';`,
    seed: 'casual.seed(0);',
    seedFunction: 'export const seedMocks = (seed: number) => casual.seed(seed);',
};

class FakerMockValueGenerator implements MockValueGenerator {
    dynamicValues: boolean;

    constructor(opts: MockValueGeneratorOptions) {
        this.dynamicValues = opts.dynamicValues;
    }

    word = () => (this.dynamicValues ? `faker.lorem.word()` : `'${faker.lorem.word()}'`);
    uuid = () => (this.dynamicValues ? `faker.datatype.uuid()` : `'${faker.datatype.uuid()}'`);
    boolean = () => (this.dynamicValues ? `faker.datatype.boolean()` : faker.datatype.boolean());
    integer = () =>
        this.dynamicValues
            ? `faker.datatype.number({ min: 0, max: 9999 })`
            : faker.datatype.number({ min: 0, max: 9999 });
    float = () =>
        this.dynamicValues
            ? `faker.datatype.float({ min: 0, max: 10, precision: 0.1 })`
            : faker.datatype.float({ min: 0, max: 10, precision: 0.1 });
    date = () =>
        this.dynamicValues
            ? `faker.date.past().toISOString(1, new Date(2022, 0))`
            : `'${faker.date.past(1, new Date(2022, 0)).toISOString()}'`;
    seed = (seed: number) => faker.seed(seed);
}

const fakerFunctionTokens: FunctionTokens = {
    import: `import { faker } from '@faker-js/faker';`,
    seed: 'faker.seed(0);',
    seedFunction: 'export const seedMocks = (seed: number) => faker.seed(seed);',
};

export const setupMockValueGenerator = ({
    generateLibrary,
    dynamicValues,
}: SetupMockValueGeneratorOptions): MockValueGenerator => {
    switch (generateLibrary) {
        case 'casual':
            return new CasualMockValueGenerator({ dynamicValues });
        case 'faker':
            return new FakerMockValueGenerator({ dynamicValues });
    }
};

export const setupFunctionTokens = (generateLibrary: 'casual' | 'faker'): FunctionTokens => {
    switch (generateLibrary) {
        case 'casual':
            return casualFunctionTokens;
        case 'faker':
            return fakerFunctionTokens;
    }
};
