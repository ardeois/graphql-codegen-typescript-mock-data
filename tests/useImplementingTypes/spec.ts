import { plugin } from '../../src';
import testSchema from './schema';

it('should support useImplementingTypes', async () => {
    const result = await plugin(testSchema, [], { prefix: 'mock', useImplementingTypes: true });

    expect(result).toBeDefined();

    expect(result).toContain(
        "config: overrides && overrides.hasOwnProperty('config') ? overrides.config! : mockTestAConfig() || mockTestTwoAConfig(),",
    );

    expect(result).toContain(
        "configArray: overrides && overrides.hasOwnProperty('configArray') ? overrides.configArray! : [mockTestAConfig() || mockTestTwoAConfig()],",
    );

    expect(result).toContain(
        "field: overrides && overrides.hasOwnProperty('field') ? overrides.field! : mockTestTwoAConfig(),",
    );

    expect(result).toContain(
        "action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : mockTestAction(),",
    );
    expect(result).toMatchSnapshot();
});

it(`shouldn't support useImplementingTypes`, async () => {
    const result = await plugin(testSchema, [], { prefix: 'mock' });

    expect(result).toBeDefined();

    expect(result).toContain(
        "config: overrides && overrides.hasOwnProperty('config') ? overrides.config! : mockAConfig(),",
    );

    expect(result).toMatchSnapshot();
});

it(`support useImplementingTypes with fieldGeneration prop`, async () => {
    let result = await plugin(testSchema, [], {
        prefix: 'mock',
        useImplementingTypes: true,
        fieldGeneration: {
            A: { str: 'email' },
        },
    });
    expect(result).toBeDefined();

    expect(result).toContain(
        "str: overrides && overrides.hasOwnProperty('str') ? overrides.str! : 'Krystel.Farrell@Frederique.biz'",
    );

    expect(result).toContain(
        "config: overrides && overrides.hasOwnProperty('config') ? overrides.config! : mockTestAConfig() || mockTestTwoAConfig(),",
    );

    result = await plugin(testSchema, [], {
        prefix: 'mock',
        useImplementingTypes: true,
        fieldGeneration: {
            A: { config: 'email' },
        },
    });
    expect(result).toBeDefined();

    expect(result).toContain("str: overrides && overrides.hasOwnProperty('str') ? overrides.str! : 'ea'");

    expect(result).toContain(
        "config: overrides && overrides.hasOwnProperty('config') ? overrides.config! : 'Karelle_Kassulke@Carolyne.io',",
    );

    expect(result).toMatchSnapshot();
});
