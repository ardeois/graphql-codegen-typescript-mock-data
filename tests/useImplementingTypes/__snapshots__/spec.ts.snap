// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should support useImplementingTypes 1`] = `
"
export const mockAConfig = (overrides?: Partial<AConfig>): AConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
    };
};

export const mockField = (overrides?: Partial<Field>): Field => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
    };
};

export const mockAction = (overrides?: Partial<Action>): Action => {
    return {
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : [TestObj.Test],
    };
};

export const mockA = (overrides?: Partial<A>): A => {
    return {
        id: overrides && overrides.hasOwnProperty('id') ? overrides.id! : 'cae147b0-1c04-459e-82db-624dd87433b4',
        str: overrides && overrides.hasOwnProperty('str') ? overrides.str! : 'ea',
        obj: overrides && overrides.hasOwnProperty('obj') ? overrides.obj! : mockB(),
        config: overrides && overrides.hasOwnProperty('config') ? overrides.config! : mockTestAConfig() || mockTestTwoAConfig(),
        configArray: overrides && overrides.hasOwnProperty('configArray') ? overrides.configArray! : [mockTestAConfig() || mockTestTwoAConfig()],
        field: overrides && overrides.hasOwnProperty('field') ? overrides.field! : mockTestTwoAConfig(),
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : mockTestAction(),
    };
};

export const mockB = (overrides?: Partial<B>): B => {
    return {
        int: overrides && overrides.hasOwnProperty('int') ? overrides.int! : 696,
        flt: overrides && overrides.hasOwnProperty('flt') ? overrides.flt! : 7.55,
        bool: overrides && overrides.hasOwnProperty('bool') ? overrides.bool! : false,
    };
};

export const mockTestAConfig = (overrides?: Partial<TestAConfig>): TestAConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
        active: overrides && overrides.hasOwnProperty('active') ? overrides.active! : true,
    };
};

export const mockTestTwoAConfig = (overrides?: Partial<TestTwoAConfig>): TestTwoAConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
        username: overrides && overrides.hasOwnProperty('username') ? overrides.username! : 'et',
    };
};

export const mockTestAction = (overrides?: Partial<TestAction>): TestAction => {
    return {
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : [TestObj.Test],
        createdAt: overrides && overrides.hasOwnProperty('createdAt') ? overrides.createdAt! : 'voluptate',
    };
};
"
`;

exports[`shouldn't support useImplementingTypes 1`] = `
"
export const mockAConfig = (overrides?: Partial<AConfig>): AConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
    };
};

export const mockField = (overrides?: Partial<Field>): Field => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
    };
};

export const mockAction = (overrides?: Partial<Action>): Action => {
    return {
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : [TestObj.Test],
    };
};

export const mockA = (overrides?: Partial<A>): A => {
    return {
        id: overrides && overrides.hasOwnProperty('id') ? overrides.id! : 'cae147b0-1c04-459e-82db-624dd87433b4',
        str: overrides && overrides.hasOwnProperty('str') ? overrides.str! : 'ea',
        obj: overrides && overrides.hasOwnProperty('obj') ? overrides.obj! : mockB(),
        config: overrides && overrides.hasOwnProperty('config') ? overrides.config! : mockAConfig(),
        configArray: overrides && overrides.hasOwnProperty('configArray') ? overrides.configArray! : [mockAConfig()],
        field: overrides && overrides.hasOwnProperty('field') ? overrides.field! : mockField(),
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : mockAction(),
    };
};

export const mockB = (overrides?: Partial<B>): B => {
    return {
        int: overrides && overrides.hasOwnProperty('int') ? overrides.int! : 696,
        flt: overrides && overrides.hasOwnProperty('flt') ? overrides.flt! : 7.55,
        bool: overrides && overrides.hasOwnProperty('bool') ? overrides.bool! : false,
    };
};

export const mockTestAConfig = (overrides?: Partial<TestAConfig>): TestAConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
        active: overrides && overrides.hasOwnProperty('active') ? overrides.active! : true,
    };
};

export const mockTestTwoAConfig = (overrides?: Partial<TestTwoAConfig>): TestTwoAConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
        username: overrides && overrides.hasOwnProperty('username') ? overrides.username! : 'et',
    };
};

export const mockTestAction = (overrides?: Partial<TestAction>): TestAction => {
    return {
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : [TestObj.Test],
        createdAt: overrides && overrides.hasOwnProperty('createdAt') ? overrides.createdAt! : 'voluptate',
    };
};
"
`;

exports[`support useImplementingTypes with fieldGeneration prop 1`] = `
"
export const mockAConfig = (overrides?: Partial<AConfig>): AConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
    };
};

export const mockField = (overrides?: Partial<Field>): Field => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
    };
};

export const mockAction = (overrides?: Partial<Action>): Action => {
    return {
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : [TestObj.Test],
    };
};

export const mockA = (overrides?: Partial<A>): A => {
    return {
        id: overrides && overrides.hasOwnProperty('id') ? overrides.id! : 'cae147b0-1c04-459e-82db-624dd87433b4',
        str: overrides && overrides.hasOwnProperty('str') ? overrides.str! : 'ea',
        obj: overrides && overrides.hasOwnProperty('obj') ? overrides.obj! : mockB(),
        config: overrides && overrides.hasOwnProperty('config') ? overrides.config! : 'Karelle_Kassulke@Carolyne.io',
        configArray: overrides && overrides.hasOwnProperty('configArray') ? overrides.configArray! : [mockTestAConfig() || mockTestTwoAConfig()],
        field: overrides && overrides.hasOwnProperty('field') ? overrides.field! : mockTestTwoAConfig(),
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : mockTestAction(),
    };
};

export const mockB = (overrides?: Partial<B>): B => {
    return {
        int: overrides && overrides.hasOwnProperty('int') ? overrides.int! : 696,
        flt: overrides && overrides.hasOwnProperty('flt') ? overrides.flt! : 7.55,
        bool: overrides && overrides.hasOwnProperty('bool') ? overrides.bool! : false,
    };
};

export const mockTestAConfig = (overrides?: Partial<TestAConfig>): TestAConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
        active: overrides && overrides.hasOwnProperty('active') ? overrides.active! : true,
    };
};

export const mockTestTwoAConfig = (overrides?: Partial<TestTwoAConfig>): TestTwoAConfig => {
    return {
        testTypes: overrides && overrides.hasOwnProperty('testTypes') ? overrides.testTypes! : [TestObj.Test],
        username: overrides && overrides.hasOwnProperty('username') ? overrides.username! : 'et',
    };
};

export const mockTestAction = (overrides?: Partial<TestAction>): TestAction => {
    return {
        action: overrides && overrides.hasOwnProperty('action') ? overrides.action! : [TestObj.Test],
        createdAt: overrides && overrides.hasOwnProperty('createdAt') ? overrides.createdAt! : 'voluptate',
    };
};
"
`;
